### ISSUE TITLE
False positive: `Cannot iterate over null` 

### ISSUE THREAD

**Issue by rbairwell** [kalessil/phpinspectionsea#1386]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         |  Php Inspections (EA Ultimate) 3.0.3 |
| Language level | PHP 7.4                  |
| Inspection | [`Foreach source to iterate over`](https://github.com/kalessil/phpinspectionsea/blob/master/docs/types-compatibility.md#foreach-source-to-iterate-over) |
| Inspection message | `Cannot iterate over null` |

Current behaviour
--
The `Cannot iterate over null` inspection comes up on the "foreach" if:
* An `[iterable](https://www.php.net/manual/en/language.types.iterable.php)` pseudotype parameter (i.e. `function test(?iterable $collection = null) : void` or `function test(?iterable $collection) : void`)
* Or a class implementing `iterator` (i.e. `?Example=null`)

has been passed as nullable: even if the parameter variable has been checked if it is null (it does not happen if `?array` has been passed)

Code showing the inspection warning (on the foreach) on an iterable:
```php
<?php
declare(strict_types = 1);

function test(?iterable $collection = null) : void
{
    if ($collection === null) {
        return;
    }
    foreach ($collection as $item) {
        print $item;
    }
}
```
Example using an iterator supporting class:
```php
<?php
declare(strict_types = 1);

class Example implements Iterator  {

    public function current()
    {
    }

    public function next()
    {
    }

    public function key()
    {
    }

    public function valid()
    {
    }

    public function rewind()
    {
    }

}

function test(?Example $collection = null) : void
{
    if ($collection === null) {
        return;
    }

    foreach ($collection as $item) {
        print $item;
    }
}
```
Expected behaviour
--
No inspection warning as:
- If you remove the nullable parameter (i.e. `?iterable` becomes `iterable`) the inspection no longer highlights (even though, as far as PHP is concerned, the code is identical).
- Swapping `iterable` for `array` does NOT show the inspection.

Environment details
---
PhpStorm 2019.2
Build #PS-192.5728.108, built on July 24, 2019
Runtime version: 11.0.3+12-b304.10 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10 10.0
GC: ParNew, ConcurrentMarkSweep
Memory: 972M
Cores: 8
Registry: run.processes.with.pty=TRUE
Non-Bundled Plugins: GrepConsole, Key Promoter X, PHPUnit code coverage, String Manipulation, com.intellij.ideolog, net.seesharpsoft.intellij.plugins.csv, NEON support, com.kalessil.phpStorm.phpInspectionsUltimate, de.espend.idea.php.phpunit, de.espend.idea.php.toolbox, de.espend.idea.php.annotation, izhangzhihao.rainbow.brackets, org.psliwa.idea.composer, ru.adelf.idea.dotenv

**Comment by rbairwell** [comment]
This also happens when the property is declared nullable with a doc block type hint.
```php
<?php
declare(strict_types = 1);

/**
 * @param iterable|null $collection
 */
function test(iterable $collection = null) : void
{
    if ($collection === null) {
        return;
    }
    foreach ($collection as $item) {
        print $item;
    }
}
```
But, again, removing the `|null` from the docblock type hint makes the inspection go awya.

**Comment by kalessil** [comment]
One of the IDE related reports is https://youtrack.jetbrains.com/issue/WI-37859, but it was similar where IDE type resolver doesn't handle complex conditions.
